pilot/helpers/agents/CodeMonkey.py

```diff
@@ -296,118 +170,180 @@ def identify_file_to_change(self, code_changes_description: str, files: list[dic
         """
-        Remove common indentation from `old_code` and `new_code`.
+        Get the diff between two files.
 
-        This is useful because the LLM will sometimes indent the code blocks MORE
-        than in the original file, leading to no matches. Since we have indent
-        compensation, we can just remove any extra indent as long as we do it
-        consistently for both old and new code block.
+        This uses Python difflib to produce an unified diff, then splits
+        it into hunks that will be separately reviewed by the reviewer.
 
-        :param old_code: old code block
-        :param new_code: new code block
-        :return: tuple of (old_code, new_code) with common indentation removed
+        :param file_name: name of the file being modified
+        :param old_content: old file content
+        :param new_content: new file content
+        :return: change hunks from the unified diff
         """
-        old_lines = old_code.splitlines()
-        new_lines = new_code.splitlines()
-        indent = 0
-        while all(ol.startswith(" ") for ol in old_lines) and all(ol.startswith(" ") for ol in new_lines):
-            indent -= 1
-            old_lines = [ol[1:] for ol in old_lines]
-            new_lines = [nl[1:] for nl in new_lines]
-        return "\n".join(old_lines), "\n".join(new_lines)
+        from_name = "old_" + file_name
+        to_name = "to_" + file_name
+        from_lines = old_content.splitlines(keepends=True)
+        to_lines = new_content.splitlines(keepends=True)
+        diff_gen = unified_diff(from_lines, to_lines, fromfile=from_name, tofile=to_name)
+        diff_txt = "".join(diff_gen)
+
+        hunks = re.split(r'\n@@', diff_txt, re.MULTILINE)
+        result = []
+        for i, h in enumerate(hunks):
+            # Skip the prologue (file names)
+            if i == 0:
+                continue
+            txt = h.splitlines()
+            txt[0] = "@@" + txt[0]
+            result.append("\n".join(txt))
+        return result
 
+    def apply_diff(
+        self,
+        file_name: str,
+        old_content: str,
+        hunks: list[str],
+        fallback: str
+    ):
+        """
+        Apply the diff to the original file content.
+
+        This uses the internal `_apply_patch` method to apply the
+        approved diff hunks to the original file content.
+
+        If patch apply fails, the fallback is the full new file content
+        with all the changes applied (as if the reviewer approved eveythng).
+
+        FIXME: For verification, it also calls command-line "Patch"
+        utility to check that the internal algorithm works correctly.
+
+        :param file_name: name of the file being modified
+        :param old_content: old file content
+        :param hunks: change hunks from the unified diff
+        :param fallback: proposed new file content (with all the changes applied)
```

Comment:
maybe it's safer that we fallback to original_file and once we implement retrying of step in case there are no file changes approved (i mentioned in comment above case when content == file_content) then this fallback should be same

Reply:
also add trace